astropy/coordinates/builtin_frames/skyoffset.py

```diff
@@ -177,3 +183,20 @@ def represent_as(self, base, s='base', in_frame_units=False):
         data = super().represent_as(base, s, in_frame_units=in_frame_units)
         self._set_skyoffset_data_lon_wrap_angle(data)
         return data
+
+
+def __getattr__(name):
+    if name in __all__ and name in globals():
+        return globals()[name]
+
+    if not name.startswith('SkyOffset'):
+        raise AttributeError(f"Module {__name__!r} has no attribute {name!r}.")
+
+    if name in _skyoffset_cache:
+        return _skyoffset_cache[name]
+
+    else:
+        # builtin_frames
+        frame_name = name[9:]
+        FrameCls = getattr(builtin_frames, frame_name, None)
```

Comment:
Should we not just fail if the frame does not exist on `builtin_frames`? 

Also, possibly this cannot happen, but what about an offset frame defined for a user-defined frame? Perhaps we should look up the frame on the transformation graph instead?

Reply:
I'll point out that we in SunPy have uses cases for defining an offset frame for one of SunPy's frames, so I'll second pulling the frame class from the transformation graph.

Reply:
I don't think reading off of the frame transform graph will work generically. The `SkyOffset<X>` frames are generated dynamically, so they are only added to the graph after `make_skyoffset_cls()` is called. If I pickle such an instance and try to load it in an independent Python instance, there's no way to make sure that the class has been defined (or exists on the graph). So I agree that what is done here won't work for user-defined skyoffset frames, but it does at least work for builtin frames. But maybe I misunderstood what you were suggesting?

Reply:
Ooh, but another option could be to implement a custom `__reduce__` method on `SkyOffsetFrame` -- I'll see if that looks better.

Reply:
See #13305 for an alternate to this PR!